### ISSUE TITLE
Closing Chromely closes top-level app

### ISSUE THREAD

**Issue by lwintjen** [chromelyapps/Chromely#128]
Hello,

I'm using Chromely as a WebView in a VSTO add-in (for Powerpoint) but when I close the window (by clicking on the "X" or by doing a window.close() from the JS side), it completely closes my top-level application which is Powerpoint. Is there any solution for this use-case?

An image of my WebView within PPTX : 
![image](https://user-images.githubusercontent.com/43049559/66652129-32ba4200-ec35-11e9-8e28-667ecd75f4d5.png)

**Comment by mattkol** [comment]
@lwintjen interesting. 

This is likely due to:
https://github.com/chromelyapps/Chromely/blob/e93c564c2fcba30049ef795831747f702dc984b0/src/Common/NativeWindowBase.cs#L480

Remember Chromely is designed to be used in standalone App so we are actually exiting the process. You may need to play around with that - or remove it entirely. So many things have happened since that was put in so may be redundant now.


Please keep us posted.

Thanks.

**Comment by lwintjen** [comment]
Thanks for your quick response!

I understand why, I didn't know that Chromely was supposed to be used as a standalone App. It's the only third-party lib that I've managed to use as a WebView for my VSTO add-in though. do you have any recommendation about how I could do to change that value?

**Comment by mattkol** [comment]
If you are using the nuget packages - as it is nothing can be done.

I can make a code change by the weekend or you remove it yourself. Either way  the source has to be rebuilt.

There is no immediate plan now to make a release.

**Comment by mattkol** [comment]
Also this does not mean this is the real issue. It is likely. But can be anything. So my suggestion is you remove it. Rebuild the source. When you rebuild, nuget packages are automatically created. So you have the option to use the dlls directly or use local nuget directory.

**Comment by lwintjen** [comment]
Allright, thanks for your explanations.

I cloned the repo locally, I made the change inside NativeWindowsBase. I deleted the nugets folder and tried to rebuild ChromelySolution.sln in VS to get the packages I need which are : Chromely.Core, Chromely.CefGlue.Winapi and Chromely.CefGlue. But it didn't seem to create the nuget packages inside the nuget folder of the repo. 

I never created a Nuget Package before, is there anything special I should do/know ? Sorry if this seems a dumb question to you.

Here are some images of my process : 
![image](https://user-images.githubusercontent.com/43049559/66658338-92b6e580-ec41-11e9-96c0-076ce490f078.png)

Build succeeded 
![image](https://user-images.githubusercontent.com/43049559/66658377-a06c6b00-ec41-11e9-82ec-1399f336f3e6.png)

But still no nugets folder
![image](https://user-images.githubusercontent.com/43049559/66658409-b5e19500-ec41-11e9-91ae-da58b8b37509.png)

**Comment by mattkol** [comment]
Individual project bin folders (or subfolders)..somewhere there will be the nuget files. I am on the move right now and may not be that helpful until evening or weekend. 

If there are no nuget files you can use the dlls directly.

**Comment by lwintjen** [comment]
No worry, your messages are already very helpful (and so quick)!

Indeed, I found each nuget packages inside the bin folders, but I'm not able to build the project after installation of the packages because of Win.SubProcesses blocking the copy of the .dll files. I tried to launch the nuget.exe inside the Subprocesses folder but it didn't seem to do anything. What am I supposed to do with the Subprocesses folder?

![image](https://user-images.githubusercontent.com/43049559/66664253-78ced000-ec4c-11e9-85e5-793bbd01005a.png)

**Comment by mattkol** [comment]
This is likely because there is a process related to the application  running. Either the app itself or a subprocess. Please try kill any process on your machine related to chromely.

Ensure this by checking the Task Manager and end the processes.
![image](https://user-images.githubusercontent.com/18384207/66701401-be79b000-ecc1-11e9-8b23-09293da7547b.png)

Also I will disable the subprocess for now:

```csharp                
var config = ChromelyConfiguration
               .Create()
                .....
               .UseDefaultSubprocess(false)
                .....
````
See more info - https://github.com/chromelyapps/Chromely/wiki/Sub-Process

**Comment by lwintjen** [comment]
Indeed, I had running processes of CefGlue.Win.Subprocesses that I killed through the Task Manager.

I've made the following changes to the NativeWindowBase file:
`                
case WM.CLOSE:       
 {             
    //OnExit();
    //User32Methods.PostQuitMessage(0);
    //Environment.Exit(0);
    break;
}`

These changes allow me to close the WebView without killing/executing the top level-app (Powerpoint). I'm still encountering an issue : once the WebView is closed, Powerpoint is bugged/freezed, I can still navigate in the menus but nothing happens when I try to interact with PPTX (adding a slide, keyboard shorcuts to execute things, menu buttons, etc.). Is there anything I should do within the NativeWindowsBase to avoid that (if you have an idea) ? I hope that you understand what I said above, if not, tell me and I'll try to provide some gifs that show what's happening!

**Comment by mattkol** [comment]
You only need to comment out the Environment.Exit(0).

```csharp                
      case WM.CLOSE:
                    {
                        OnExit();
                        User32Methods.PostQuitMessage(0);
                      //  Environment.Exit(0);
                        break;
                    }
````
The issue we are dealing with here is to avoid Chromely closing the main application. 

What you are trying to do overall is embedding Chromely app in an application. This is not what Chromely is for. However like I said earlier you can customize it for that but it may take a lot of effort to get what you want - depending on how familiar you are with CEF generally. 

If this change does not help, I will suggest you start with [CefGlue](https://gitlab.com/xiliumhq/chromiumembedded/cefglue) or [CefSharp](https://github.com/cefsharp/CefSharp) as-is and then if you need more functionalities you can see what Chromely provides. CefSharp site is super active and [interactive](https://gitter.im/cefsharp/CefSharp?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge) and they may help with keyboard functionalities too.

**Comment by lwintjen** [comment]
This change helped (I thought I tried this way the first time that I modified the file but apparently not), I'm now able to close a Chromely instance without having the app freezing/bugging. But I'm not able to open a second instance of Chromely after the first one closed, is there anything we can do to achieve that ?

Thanks for your help.

**Comment by mattkol** [comment]
I can only guess here.

What you have done is likely to close Chromely browser. So we are likely back to square 1. The console app hosting the browser is likely still running. So what you have done is not likely the the ultimate solution. 

So a different approach will be to find the console process handle and kill it instead of the Environment.Exit(0). 

Is there a related process running that you can see?
 
Again all this is just from experience I don't have a full picture.

I can look at that approach this evening and will get back to you if you have not found a solution yet.

I am curious too.

**Comment by mattkol** [comment]
Please do some research around:

````csharp
var process = Process.GetCurrentProcess()
process.Kill()
````

Something like that will replace the Exit(0) command.

**Please make sure it is the right process. You can also search process by name.**

**Comment by lwintjen** [comment]
Indeed, I had ghost processes running so I added this :

```c#
        try
            {
              foreach (Process proc in Process.GetProcessesByName("Chromely.CefGlue.Win.Subprocess"))
              {
                proc.Kill();
              }
            }
        catch (Exception ex)
            {
              MessageBox.Show(ex.Message);
            }
```
With your suggestion, it just closed the ghost process and also the application (so killing the whole process : WebView + PPTX).

Here are the processes running when launching PPTX :
![image](https://user-images.githubusercontent.com/43049559/66769769-39a3b780-eeb6-11e9-8f87-c479b45f0c7e.png)

Then after opening Chromely within PPTX, I have 2 threads of CefGlue running :
![image](https://user-images.githubusercontent.com/43049559/66769794-49230080-eeb6-11e9-980c-040ef2296c72.png)

After closing Chromely GUI, the 2  threads are killed successfully then I have other threads that are killed (I guess it's normal):
![image](https://user-images.githubusercontent.com/43049559/66769812-563fef80-eeb6-11e9-8dce-af369ef36ecd.png)
![image](https://user-images.githubusercontent.com/43049559/66769841-6ce64680-eeb6-11e9-8a43-598acfdb31c7.png)
![image](https://user-images.githubusercontent.com/43049559/66769806-517b3b80-eeb6-11e9-9dc1-dccf92e4e50a.png)

So from there, everything seems normal but I still cannot reopen an instance of Chromely/CefGlue.

I have thought about a workaround, tell me if this would be possible : 
 - Whenever we open the top-level application, in this scenario Powerpoint, we also run a CefSharp or CefGlue process but no GUI (Chromely) is visible. And then when we try to open the WebView, we bound the running webbrower process to Chromely and we make the GUI visible. 

We would thus always have a browser running but the user would be able to use his application normally without even knowing that there's a browser that is waiting to be executed. Does that make sense ? Or do you have another workaround in your mind? 

We could also try to make the browser a process on his own instead of a thread of PPTX (is this doable?).

**Comment by mattkol** [comment]
A sub-process lifetime is based on the parent's lifetime. Killing a sub-process that way I will guide against that.

The process I am looking for is the main process. For Chromely this is a Console App. From your diagram it looks like you are already doing pure embedding. Reason why you are having the original issue. So killing the main process will likely kill the PPTX app too.

At this stage I will revert to my original suggestion s.
1. Comment out the Exit(o) code.
2. Remove sub-process as described here..
https://github.com/chromelyapps/Chromely/issues/128#issuecomment-541320661

Thanks.

**Comment by lwintjen** [comment]
I don't seem to find any other process related to CEF. The two sub-processes are the two I found related to CEF/CefSharp/CefGlue. And I don't find any process that is not a sub-process to PPTX.
![image](https://user-images.githubusercontent.com/43049559/66775417-3283a600-eec4-11e9-86a3-b09d70a3247c.png)


Well, I added the changes to the existing one, so exit(o) was already commented out. But I tried what you said above and I cannot access Chromely when I remove the sub-process, it makes it buggy (whenever I try to access the process, it minimizes it so it blocks the access to it) with the following configuration :
```
      ChromelyConfiguration config = ChromelyConfiguration
                                    .Create()
                                    .UseDefaultSubprocess(false)
                                    .WithDebuggingMode(true)
                                    .WithHostMode(Chromely.Core.Host.WindowState.Normal)
                                    .WithHostTitle("Wooclap")
                                    .WithHostIconFile("chromely.ico")
                                    .WithHostBounds(1200, 700)
                                    .UseDefaultResourceSchemeHandler("local", string.Empty)
                                    .UseDefaultHttpSchemeHandler("http", "localhost:3000")
                                    .UseDefaultHttpSchemeHandler("wooclap", "api.com")
                                    .WithStartUrl(startUrl);
```

To make sure that you didn't misunderstood, here is what I modified :
```
 case WM.CLOSE:
                    {
                      OnExit();
                      User32Methods.PostQuitMessage(0);
                      //Environment.Exit(0);
                      try
                      {
                        foreach (Process proc in Process.GetProcessesByName("Chromely.CefGlue.Win.Subprocess"))
                        {
                          proc.Kill();
                        }
                      }
                      catch (Exception ex)
                      {
                        MessageBox.Show(ex.Message);
                      }
                      break;
                    }
```

**Comment by mattkol** [comment]
I think we are going in cycles now. Can you please show me a pseudo codo of how Chromely was launched? Does not seem like a separate isolated process to me. Is it like we do in the demos..with Program.cs? or a class is created?

**Comment by lwintjen** [comment]
Chromely is implemented like described in the wiki. The only class created is a DemoController to communicate between my WebView and C#.
```
private void button1_Click(object sender, RibbonControlEventArgs e)
    {
      
      string startUrl = "http://localhost:3000/";

      ChromelyConfiguration config = ChromelyConfiguration
                                    .Create()
                                    .UseDefaultSubprocess(false)
                                    .WithDebuggingMode(true)
                                    .WithHostMode(Chromely.Core.Host.WindowState.Normal)
                                    .WithHostTitle("Wooclap")
                                    .WithHostIconFile("chromely.ico")
                                    .WithHostBounds(1200, 700)
                                    .UseDefaultResourceSchemeHandler("local", string.Empty)
                                    .UseDefaultHttpSchemeHandler("http", "localhost:3000")
                                    .UseDefaultHttpSchemeHandler("wooclap", "api.com")
                                    .WithStartUrl(startUrl);

      using (var window = ChromelyWindow.Create(config))
      {

        window.RegisterServiceAssembly(Assembly.GetExecutingAssembly());

        // 2. Register external assembly with file name:
        // string serviceAssemblyFile = @"C:\ChromelyDlls\Chromely.Service.Demo.dll";
        // window.RegisterServiceAssembly(serviceAssemblyFile);

        // 3. Register external assemblies with list of filenames:
        // string serviceAssemblyFile1 = @"C:\ChromelyDlls\Chromely.Service.Demo.dll";
        // List<string> filenames = new List<string>();
        // filenames.Add(serviceAssemblyFile1);
        // window.RegisterServiceAssemblies(filenames);

        // 4. Register external assemblies directory:
        // string serviceAssembliesFolder = @"C:\ChromelyDlls";
        // window.RegisterServiceAssemblies(serviceAssembliesFolder);

        // Scan assemblies for Controller routes 
        window.ScanAssemblies();

        window.Run(new string[] { });

      }
```

**Comment by mattkol** [comment]
Interesting...I never really tried this before. Now I get the real picture.

The application process..main process is tied to Chromely process this way.

I personally do not think it will work as you expect..and frankly not advisable. Unless you want to have a single browser. Once it launches you show....consequently you hide or show the browser. Which won't be a bad idea..if that meets your success criteria.

My approach will be to isolate it. You can use Process.Start with arguments to launch it.

Because I am curious too I will show you what I mean...later this evening..I will try...

So to recap two options;
1: leave as is and show and hide as needed..
2. Create an isolated procees and pass arguments.

**Comment by lwintjen** [comment]
Exactly!

I'll try your approach and see how it goes. Otherwise, I'll do the other one. Is there any option in Chromely to programmatically hide/show the GUI? (I don't think that I have seen any option of that kind in the wiki, maybe setting dynamically the width/height to 0/0 or 1200/700 would do the trick).

By isolating, you mean, that we set Chromely as a process on his own and not a sub-process (tied) to the main process. As I describe above ?

**Comment by mattkol** [comment]
On show and hide, there is no functionality like that in the project, but it can be done. You can look into sending/posting message using the [Window handle](https://github.com/chromelyapps/Chromely/blob/7a7abd8fc175b6d3ffda9312b9105ea5fa50efe1/src/Chromely.Core/Host/IChromelyWindow.cs#L32).

Something like:
````
ShowWindow(Handle, SW_HIDE);
````

On separate process option  .. I mean create a normal  chromely application (exe) and launch it. The challenge will be how to pass parameters, but you can see how to pass json - serializing/de-serializing.

.. launching:
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.start?view=netframework-4.8
```
        private void button1_Click(object sender, RibbonControlEventArgs e)
        {
            try
            {
                using (Process myProcess = new Process())
                {
                    myProcess.StartInfo.UseShellExecute = false;
                    // You can start any process, HelloWorld is a do-nothing example.
                    myProcess.StartInfo.FileName = "ChromelyApp.exe";
                    myProcess.StartInfo.CreateNoWindow = true;
                    myProcess.Start();
                    // This code assumes the process you are starting will terminate itself. 
                    // Given that is is started without a window so you cannot terminate it 
                    // on the desktop, it must terminate itself or you can do it programmatically
                    // from this application using the Kill method.
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
          
```
**Note that using this option will require no changes to the source code. Environment.Exit(0) will stay**

I just tested this option and it works.

**Comment by lwintjen** [comment]
Thanks for your test!

Do you know where I can find the `.exe` file to execute because it doesn't execute anything at the moment. I think that the file to execute doesn't exist in the working directory. 
I'll try to communicate the parameters to the process.

**Comment by mattkol** [comment]
The exe will come from the console app you build. Assuming you are using the [cefglue demo](https://github.com/chromelyapps/demo-projects/tree/master/regular-chromely/Chromely.CefGlue.Winapi.Demo) - when you build there will be an exe like "cefglue_winapi_demo.exe" you will copy everything in this folder to your application directory.

If you are using a custom build, you will follow the same pattern.

Then:

````
 myProcess.StartInfo.FileName = "cefglue_winapi_demo.exe";
````

If show/hide will work for you, that may be a better option with dealing with parameters, but keyboard issue may crop up.

Even though I know these approaches will work, I still do not think it is the right way to go. Launching/re-launching, show/hide may not be an issue, but you may run into other issues, like keyboard issues .. if you expect it to be modal, the launch/re-launch will not be.

This is really up to you, but I will go with real embedding, It may look difficult at the begining but if you spend a few days on it, it may be worth it.

**If you want to try embedding:**

https://github.com/cefsharp/CefSharp.MinimalExample

For this I will stay with CefSharp, because you may find answers faster than CefGlue. Try running CefSharp WinForm, WPF samples. And work backwards stripping off what you do not need. Once it works, you can then see how you can include Chromely functionalities.

Please ensure you pick the right version that Chromely supports if you need Chromely support.

**Comment by lwintjen** [comment]
Thanks for these informations.

I've tried to dynamically hide and show the window using `showWindow()`. It's working : we can click on the button from the browser that hides the iframe and we can reopen it by clicking on the same button. Here's my process : I'm launching Chromely at the beginning at the start of the application but I don't run it. The first time the user tries to open the iframe, I run Chromely then instead of closing Chromely and exiting the process I do :
`User32Methods.ShowWindow(Globals.ThisAddIn.WooclapWindow.Handle, ShowWindowCommands.SW_HIDE);`

And if the user tries to reopen it then I simply use `ShowWindowCommands.SH_SHOW` on the same instance of Chromely. `WooclapWindow` is the instance of Chromely that I launched on the start of the application.

The only issue I'm currently facing is when we close Chromely from the border menu (by clicking on the "X") as we cannot reopen another iframe. I tried to change the code in NativeWindowBase :
```
case WM.CLOSE:
                    {
                        //OnExit();
                        //User32Methods.PostQuitMessage(0);
                        User32Methods.ShowWindow(Handle, ShowWindowCommands.SW_HIDE);
                        //Environment.Exit(0);
                        break;
                    }
```

I tried to comment and uncomment the first two lines of this switch case but the result is the same. Is there anything else you do (outside of this block of code) when the user closes the GUI that might block to run another instance ?

**Comment by mattkol** [comment]
You did the right thing.

What is left is to return from the WM.CLOSE. Instead of break, you want to stop the message propagation. I don't have the exact answer for that but you can check online. If you don't find a solution by evening.. I can check in the evening.

**Comment by lwintjen** [comment]
I changed the break for a return `return IntPtr.Zero;` still doesn't fix it. I'll still try to find a fix in the meantime!
